libs/wire-api/test/unit/Test/Wire/API/Golden/Runner.hs

```diff
@@ -34,37 +35,87 @@ import Data.ProtoLens.Message (Message)
 import Data.ProtoLens.TextFormat (pprintMessage, readMessage)
 import qualified Data.Text.Lazy.IO as LText
 import Imports
+import Polysemy (Embed, Members, Sem, embed, runM)
+import Polysemy.Error (Error, runError, throw)
 import Test.Tasty.HUnit
 import Text.PrettyPrint (render)
 import Type.Reflection (typeRep)
 import Wire.API.ServantProto
 
+data TestOutcomeError a
+  = FileDoesNotExist FilePath
+  | FailedToDecode String
+  | AesonValueMismatch
+      Value
+      -- ^ Expected value
+      Value
+      -- ^ Actual value
+      FilePath
+      -- ^ The path to the file
+  | FailedToParse a (Result a) FilePath
+
+instance (Typeable a, Show a) => Show (TestOutcomeError a) where
+  show = \case
+    FileDoesNotExist path -> "Golden JSON file " <> show path <> " does not exist"
+    FailedToDecode err -> "Failed to decode: " <> err
+    AesonValueMismatch expected actual path ->
+      show (typeRep @a) <> ": ToJSON should match golden file: " <> show path
+        <> "\n"
+        <> "expected: "
+        <> show expected
+        <> "\n"
+        <> " but got: "
+        <> show actual
+    FailedToParse obj result path ->
+      (show (typeRep @a) <> ": FromJSON of " <> show path <> " should match object")
+        <> "\n"
+        <> "expected: "
+        <> show (Success obj)
+        <> "\n"
+        <> " but got: "
+        <> show result
+
 testObjects :: forall a. (Typeable a, ToJSON a, FromJSON a, Eq a, Show a) => [(a, FilePath)] -> IO ()
 testObjects objs = do
-  allFilesExist <- and <$> traverse (uncurry testObject) objs
-  assertBool "Some golden JSON files do not exist" allFilesExist
+  allResults <- traverse (run . uncurry testObject) objs
+  let errorResults = foldMap (either pure (const [])) allResults
+      errorMsgs = intercalate "\n" . fmap show $ errorResults
+  assertBool errorMsgs . null $ errorResults
```

Comment:
Instead of traversing and collecting errors, could we make each one these a `testCase`? That way we don't have to invent any of this.

Reply:
Done!

As a consequence, the output of `stack test wire-api` is much more verbose and (correctly) reports 3549 passing tests (compared to ~570 tests before turning each object into a test case).